pkg/jobWorker/job.go

```diff
@@ -0,0 +1,374 @@
+					job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error unmounting /proc - %w\n", err))
+				}
+			}
+		}
+	}()
+
+	if err = ns.MountProc(); err != nil {
+		unmount <- true
+		return fmt.Errorf("Error mounting /proc - %w\n", err)
+	}
+
+	log.Printf("starting job:%s, cmd:%s", job, cmd.String())
+	if err = cmd.Start(); err != nil {
+		cleanCGroup <- true
+		unmount <- true
+		return fmt.Errorf("error starting command: %w", err)
+	}
+	job.cmd = cmd
+	job.isStarted = true
+
+	// run the command in a Goroutine so that Start can return immediately
+	go func() {
+		// Use cmd.Process.Wait() instead of cmd.Wait() since cmd.Wait() is not thread safe
+		// and we do not want to hold the mutex while waiting for the process to exit.
+		// So instead we use cmd.Process.Wait() and store the result in j.processState to mimic what
+		// cmd.Wait would do.
+		// This prevents concurrency issues when a user calls Start(), the command quickly exits (updating the
+		// process state), and the user invokes Status().
+		processState, err := job.cmd.Process.Wait()
+		job.processState = processState
+
+		job.mutex.Lock()
+		defer job.mutex.Unlock()
+
+		// at this stage job in completed (successfully or not we can detect from checking job.exitReason and isTerminated )
+		job.isCompleted = true
+
+		defer func() {
+			cleanCGroup <- true
+			unmount <- true
+		}()
+
+		if err != nil {
+			job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error running command: %w\n", err))
+		}
+
+		if err == nil && !job.processState.Success() {
+			job.exitReason = errors.Join(job.exitReason, &exec.ExitError{ProcessState: job.processState})
+		}
+
+		// close the output, so that any readers of the output know the process has exited and will no longer
+		// block waiting for new output
+		if err = job.output.Close(); err != nil {
+			job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing output: %w\n", err))
+		}
+	}()
+
+	return nil
+}
+
+// Status returns the current Status of the Job.
+func (job *Job) Status() *JobStatus {
+	job.mutex.Lock()
+	defer job.mutex.Unlock()
+
+	if !job.isStarted {
+		return &JobStatus{
+			State:    JOB_STATUS_NOT_STARTED,
+			ExitCode: -1,
+		}
+	}
+
+	if job.isStarted && !job.isCompleted {
+		return &JobStatus{
+			State:    JOB_STATUS_RUNNING,
+			ExitCode: -1,
+		}
+	}
+
+	if job.isTerminated {
+		return &JobStatus{
+			State:      JOB_STATUS_TERMINATED,
+			ExitCode:   job.processState.ExitCode(),
+			ExitReason: job.exitReason.Error(),
+		}
+	}
+
+	return &JobStatus{
+		State:      JOB_STATUS_COMPLETED,
+		ExitCode:   job.processState.ExitCode(),
+		ExitReason: job.exitReason.Error(),
+	}
+}
+
+// Stream returns an OutputReadCloser (implements io.ReadCloser)  that streams the combined stdout and stderr of the Job.
+func (job *Job) Stream() io.ReadCloser {
+	log.Printf("get job stream:%s", job)
+	return NewOutputReadCloser(job.output)
+}
+
+func (job *Job) Stop() error {
```

Comment:
nit: what happens if `Stop` is called after the cmd is terminated

Reply:
@greedy52 returns ErrJobAlreadyStopped.
<img width="780" alt="Screenshot 2024-09-05 at 11 34 39 AM" src="https://github.com/user-attachments/assets/2a8a7c85-3fe8-4066-8719-f41d3eb15cf9">

Reply:
do you need lock for `job.isTerminated` and `job.isCompleted`?

Reply:
Sorry, incorrect screenshot - it was moved after mutex.Lock
<img width="724" alt="Screenshot 2024-09-05 at 3 27 37 PM" src="https://github.com/user-attachments/assets/e4f7405c-97dc-4a44-b6d5-583d8542ea0d">